File: leetcode/easy/155_min_stack.js
// Design a stack that supports push, pop, top, and retrieving the minimum element in constant time.

// push(x) -- Push element x onto stack.
// pop() -- Removes the element on top of the stack.
// top() -- Get the top element.
// getMin() -- Retrieve the minimum element in the stack.

// To receive something in constant time, we need to use a data structure to store certain values.

var MinStack = function() {
  this.minStack = [];
  this.stack = [];
};

/** 
 * @param {number} x
 * @return {void}
 */
MinStack.prototype.push = function(x) {
  if (this.minStack.length === 0 || x <= this.minStack[this.minStack.length - 1]) {
    this.minStack.push(x);
  }
  this.stack.push(x);
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function() {
  let item = this.stack.pop();
  if (item === this.minStack[this.minStack.length - 1]) {
    this.minStack.pop();
  }
};

/**
 * @return {number}
 */
MinStack.prototype.top = function() {
  return this.stack[this.stack.length - 1];
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function() {
  return this.minStack[this.minStack.length - 1]
};

 

// Example:

// MinStack minStack = new MinStack();
// minStack.push(-2);
// minStack.push(0);
// minStack.push(-3);
// minStack.getMin();   --> Returns -3.
// minStack.pop();
// minStack.top();      --> Returns 0.
// minStack.getMin();   --> Returns -2.

File: python/array/155_minstack.py
# Design a stack that supports push, pop, top, and retrieving the minimum element in constant time.

# push(x) -- Push element x onto stack.
# pop() -- Removes the element on top of the stack.
# top() -- Get the top element.
# getMin() -- Retrieve the minimum element in the stack.
 

# Example:

# MinStack minStack = new MinStack();
# minStack.push(-2);
# minStack.push(0);
# minStack.push(-3);
# minStack.getMin();   --> Returns -3.
# minStack.pop();
# minStack.top();      --> Returns 0.
# minStack.getMin();   --> Returns -2.

# class MinStack:
  
#     def __init__(self):
#         """
#         initialize your data structure here.
#         """
#         self.stack = []
        

#     def push(self, x: int) -> None:
      
#         if not self.stack:
#           self.stack.append((x, x))
#         else:
#           curr_min = min(x, self.stack[-1][1])
#           self.stack.append((x, curr_min))

#     def pop(self) -> None:
#         if self.stack:
#           self.stack.pop()

#     def top(self) -> int:
#         if self.stack:
#           return self.stack[-1][0]
#         else:
#           return None

#     def getMin(self) -> int:
#         if self.stack:
#           return self.stack[-1][1]


# # Your MinStack object will be instantiated and called as such:
# # obj = MinStack()
# # obj.push(x)
# # obj.pop()
# # param_3 = obj.top()
# # param_4 = obj.getMin()

# Time : O(1) | O(n)